mobile_verifier/src/reward_shares.rs

```diff
@@ -197,6 +204,112 @@ impl MapperShares {
+        for (payer, total_dcs) in payer_shares {
+            let service_provider = Self::payer_key_to_service_provider(&payer, client).await?;
+            sp_shares.shares.push(ServiceProviderDataSession {
+                service_provider,
+                total_dcs: Decimal::from(total_dcs),
+            })
+        }
+        Ok(sp_shares)
+    }
+
+    fn total_dc(&self) -> Decimal {
+        self.shares.iter().map(|v| v.total_dcs).sum()
+    }
+
+    pub fn get_scheduled_tokens_for_service_providers(&self, duration: Duration) -> Decimal {
+        get_total_scheduled_tokens(duration) * SERVICE_PROVIDER_PERCENT
+    }
+
+    pub fn rewards_per_share(
+        &self,
+        total_sp_rewards: Decimal,
+        mobile_bone_price: Decimal,
+    ) -> anyhow::Result<Decimal> {
+        let total_sp_dc = self.total_dc();
+        let total_sp_rewards_used = dc_to_mobile_bones(total_sp_dc, mobile_bone_price);
+        let capped_sp_rewards_used =
+            Self::maybe_cap_service_provider_rewards(total_sp_rewards_used, total_sp_rewards);
+        Ok(Self::calc_rewards_per_share(
+            capped_sp_rewards_used,
+            total_sp_dc,
+        ))
+    }
+
+    pub fn into_service_provider_rewards(
+        self,
+        reward_period: &'_ Range<DateTime<Utc>>,
+        reward_per_share: Decimal,
+    ) -> impl Iterator<Item = proto::MobileRewardShare> + '_ {
+        self.shares
+            .into_iter()
+            .map(move |share| proto::ServiceProviderReward {
+                service_provider_id: share.service_provider as i32,
+                amount: (share.total_dcs * reward_per_share)
+                    .round_dp_with_strategy(0, RoundingStrategy::ToZero)
+                    .to_u64()
+                    .unwrap_or(0),
+            })
+            .filter(|service_provider_reward| service_provider_reward.amount > 0)
+            .map(|service_provider_reward| proto::MobileRewardShare {
+                start_period: reward_period.start.encode_timestamp(),
+                end_period: reward_period.end.encode_timestamp(),
+                reward: Some(ProtoReward::ServiceProviderReward(service_provider_reward)),
+            })
+    }
+
+    fn maybe_cap_service_provider_rewards(
+        total_sp_rewards_used: Decimal,
+        total_sp_rewards: Decimal,
+    ) -> Decimal {
+        match total_sp_rewards_used <= total_sp_rewards {
+            true => total_sp_rewards_used,
+            false => total_sp_rewards,
+        }
+    }
+
+    fn calc_rewards_per_share(total_rewards: Decimal, total_shares: Decimal) -> Decimal {
+        if total_shares > Decimal::ZERO {
+            (total_rewards / total_shares)
+                .round_dp_with_strategy(DEFAULT_PREC, RoundingStrategy::MidpointNearestEven)
+        } else {
+            Decimal::ZERO
+        }
+    }
+
+    async fn payer_key_to_service_provider(
+        payer: &str,
+        client: &dyn CarrierServiceVerifier<Error = ClientError>,
```

Comment:
isn't this easier to read as &impl CarrierServiceVerifier<Error = ClientError>

Reply:
addressed here: e4b9890f0b20b7272b15c56710ae389dde8d573a